include/alpaka/pltf/PltfOmp5.hpp

```diff
@@ -0,0 +1,105 @@
+#endif
+
+#include <alpaka/pltf/Traits.hpp>
+#include <alpaka/dev/DevOmp5.hpp>
+#include <alpaka/core/Concepts.hpp>
+
+#include <sstream>
+#include <vector>
+#include <limits>
+
+namespace alpaka
+{
+    namespace pltf
+    {
+        //#############################################################################
+        //! The CPU device platform.
+        class PltfOmp5 :
+            public concepts::Implements<ConceptPltf, PltfOmp5>
+        {
+        public:
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST PltfOmp5() = delete;
+        };
+    }
+
+    namespace dev
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The CPU device device type trait specialization.
+            template<>
+            struct DevType<
+                pltf::PltfOmp5>
+            {
+                using type = dev::DevOmp5;
+            };
+        }
+    }
+    namespace pltf
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The CPU platform device count get trait specialization.
+            template<>
+            struct GetDevCount<
+                pltf::PltfOmp5>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getDevCount()
+                -> std::size_t
+                {
+                    ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                    const std::size_t count = static_cast<std::size_t>(::omp_get_num_devices());
+                    // runtime will report zero devices if the host is the target (or without offloading)
+                    return count > 0 ? count : 1;
+                }
+            };
+
+            //#############################################################################
+            //! The CPU platform device get trait specialization.
+            template<>
+            struct GetDevByIdx<
+                pltf::PltfOmp5>
+            {
+                //-----------------------------------------------------------------------------
+                //! \param devIdx device id, less than GetDevCount, will be set to omp_get_initial_device() otherwise
+                ALPAKA_FN_HOST static auto getDevByIdx(
+                    std::size_t devIdx)
+                -> dev::DevOmp5
+                {
+                    ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                    std::size_t const devCount(static_cast<std::size_t>(::omp_get_num_devices()));
+                    int devIdxOmp5 = static_cast<int>(devIdx);
+                    if(devIdx >= devCount)
+                    { // devIdx param must be unsigned, take take this case to use the initial device
```

Comment:
"take take"?
The other platforms throw an exception in the case of an invalid `devIdx`. We should do this here as well.

Reply:
OpenMP target has target devices with `0 <= devIdxOmp5 < devCount`. In addition, there is an "initial device", which has a `devIdxOmp5` which is implementation defined, all compilers I saw have `devIdxOmp5 < 0`. The trait's parameter is size_t, so we cannot map this directly, so I decided to use `devIdx > devCount`. Apparently, they want change this in OpenMP 5.1 to mandate `initialDevice == devCount`.

Reply:
`devIdx` and `devCount` are in general totally unrelated to `devIdxOmp5`. We could have an arbitrary mapping between those.
I would imagine that we always artificially add one to the value returned by `omp_get_num_devices()` to account for the default-device when giving the user the number of available devices. `devIdx == omp_get_num_devices()` would request this default device. All other indices would directly map to the `devIdxOmp5` and `devIdx > omp_get_num_devices()` would result in the exception.
We could also change this mapping to make `devIdx == 0` the default-device and offset all others by one.

Reply:
Including the initial device in `devCount` would break performance of programs which map ranks to device in a wrap-around fashion or which try to use all devices reported by alpaka. The normal use case would be only use the actual target devices and not the host, i.e. `0 <= devIdx < devCount`. Only there are not target devices available it makes sense include the host and thus have `devCount = 1` instead of `0`, merge OpenMP behavior with normal alpaka-backend behavior. There is no alpaka interface for using the initial device/host through the backend. Doing this is purely an OpenMP traget thing, thus only this case should deviate from normal alpaka behavior. If the user code is not aware which backend is being used, it should be able expect the same behavior from all backends, as long as it uses the backends properly through the interface.

We could limit this to using the initial device only if `devIdx == omp_get_num_devices()`, because this can be elegantly checked, and throw on other values.

This would of course still lead to the same unexpected behavior if user code were to iterate the devices like this:
```c++
try
{
 for(int a; ;++a)
 {
  getDevice(a);
  //...
 }
}
catch (out_of_range){}
```
but I think we do not need to cater to bad style.

Reply:
> We could limit this to using the initial device only if devIdx == omp_get_num_devices(), because this can be elegantly checked, and throw on other values.

Sounds good

Reply:
https://github.com/alpaka-group/alpaka/pull/1126/commits/2db34b82d4586a5ea853005aba532e9b5ddfccf2

In addition to the above condition, it also uses the initial device if `devCount == 0 && devIdx == 1`, that is when `devIdx` is equal to the return value of `getDevCount()`, which is this case differs from `omp_get_num_devices()`.